### PR TITLE
Fix name conflicts caused by the `SystemParam` and `WorldQuery` macros

### LINKED ISSUE #1727: derive(SystemParam) fails when field type name is that struct's name + "State"
## Bevy version

785aad92

## What you did

```rust
#[derive(SystemParam)]
pub struct FooParam<'a> {
    state: Res<'a, FooParamState>,
}

struct FooParamState;
```

## What you expected to happen

Compile successfully.

## What actually happened

Got this error:

```
the name `FooParamState` is defined multiple times
`FooParamState` must be defined only once in the type namespace of this module
```

## Additional information

I'm pretty sure the bug is here:

https://github.com/bevyengine/bevy/blob/45b2db70705da24a89426e6b6e77d603a3983025/crates/bevy_ecs/macros/src/lib.rs#L374

### LINKED ISSUE #8010: WorldQuery derive State and Fetch types are still accidentally nameable
## Bevy version

Latest commit as of writing ([`7d9cb1c`](https://github.com/bevyengine/bevy/commit/7d9cb1c4ab210595c5228af0ed4ec7d095241db5))

## What you did

```rust
#[derive(WorldQuery)]
pub struct Client<S: ClientState> {
    pub state: &'static S,
    // ...
}

pub trait ClientState: Component {}
```

## What went wrong

```rust
error[E0404]: expected trait, found struct `ClientState`
  --> crates/example/src/client.rs:11:22
   |
11 | pub struct Client<S: ClientState> {
   |                      ^^^^^^^^^^^ not a trait
...
35 | pub trait ClientState: Component {
   |           ----------- you might have meant to refer to this trait
   |
help: consider importing this trait instead
   |
1  | use crate::client::ClientState;
   |
```

## Additional information

To workaround it currently, we can prepend `self::` to the struct bound:
```rust
#[derive(WorldQuery)]
pub struct Client<S: self::ClientState> {
    pub state: &'static S,
    // ...
}
```

### PR DESCRIPTION
**Pull request by JoJoJet**
# Objective

Fix #1727
Fix #8010

Meta types generated by the `SystemParam` and `WorldQuery` derive macros can conflict with user-defined types if they happen to have the same name.

## Solution

In order to check if an identifier would conflict with user-defined types, we can just search the original `TokenStream` passed to the macro to see if it contains the identifier (since the meta types are defined in an anonymous scope, it's only possible for them to conflict with the struct definition itself). When generating an identifier for meta types, we can simply check if it would conflict, and then add additional characters to the name until it no longer conflicts with anything.

The `WorldQuery` "Item" and read-only structs are a part of a module's public API, and thus it is intended for them to conflict with user-defined types.

### CODE REVIEW

File: crates/bevy_ecs/src/query/fetch.rs [hunk 1]
```diff
@@ -1388,3 +1388,22 @@ unsafe impl<Q: WorldQuery> WorldQuery for NopWorldQuery<Q> {
 
 /// SAFETY: `NopFetch` never accesses any data
 unsafe impl<Q: WorldQuery> ReadOnlyWorldQuery for NopWorldQuery<Q> {}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate as bevy_ecs;
+
+    // regression test for https://github.com/bevyengine/bevy/issues/8010.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/fetch.rs:1397]: I'd really like it if this was moved into a discrete test with a name. It's tricky to grasp the structure here, and seems easy to revert out of confusion.

File: crates/bevy_macro_utils/src/lib.rs [hunk 2]
```diff
@@ -108,6 +108,41 @@ impl BevyManifest {
     }
 }
 
+/// Finds an identifier that will not conflict with the specified set of tokens.
+/// If the identifier is present in `haystack`, extra characters will be added
+/// to it until it no longer conflicts with anything.
+///
+/// Note that the returned identifier can still conflict in niche cases,
+/// such as if an identifier in `haystack` is hidden behind an un-expanded macro.
+pub fn ensure_no_collision(value: Ident, haystack: TokenStream) -> Ident {
```
Comment (alice-i-cecile) [inline crates/bevy_macro_utils/src/lib.rs:117]: I love that this is reusable :D

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Test needs some cleanup, then LGTM.

File: crates/bevy_ecs/src/query/fetch.rs (hunk 1 above)
Comment (JoJoJet) [reply crates/bevy_ecs/src/query/fetch.rs:1397]: Fair enough, I'm just not aware of any "runnable" code I could include in a test that would be relevant here.

Comment (alice-i-cecile) [reply crates/bevy_ecs/src/query/fetch.rs:1397]: Hmm. You should be able to define types within functions just fine, based on a quick playground experiment. `assert_is_system` may also be nice to toss on though.

File: crates/bevy_macro_utils/src/lib.rs [hunk 3]
```diff
@@ -108,6 +108,48 @@ impl BevyManifest {
...
+    // Collect all the identifiers in `haystack` into a set.
+    let idents = {
+        // List of token streams that will be visited in future loop iterations.
+        let mut unvisited = vec![haystack];
+        // Identifiers we have found while searching tokens.
+        let mut found = HashSet::new();
```
Comment (james7132) [inline crates/bevy_macro_utils/src/lib.rs:123]: This still requires us to rerun the TokenStream multiple times if `ensure_no_collision` called multiple times in the same macro. IMO we shouldn't promote the use of this in `bevy_macro_utils` until we retain this `HashSet` between calls to this function.

File: crates/bevy_macro_utils/src/lib.rs [hunk 4]
```diff
@@ -8,10 +8,10 @@ pub use attrs::*;
 pub use shape::*;
 pub use symbol::*;
 
-use proc_macro::TokenStream;
+use proc_macro::{TokenStream, TokenTree};
 use quote::{quote, quote_spanned};
-use std::{env, path::PathBuf};
-use syn::spanned::Spanned;
+use std::{collections::HashSet, env, path::PathBuf};
```
Comment (james7132) [inline crates/bevy_macro_utils/src/lib.rs:13]: We probably can't use `bevy_utils::HashSet` here, but we can use `rustc-hash`'s hasher.

**Review by james7132 — CHANGES_REQUESTED** [review]

File: crates/bevy_macro_utils/src/lib.rs (hunk 3 above)
Comment (JoJoJet) [reply crates/bevy_macro_utils/src/lib.rs:123]: Proc macros have so much overhead, I highly doubt that micro-optimizations like this are going to matter.

**Review by james7132 — APPROVED** [review]

**Pull request merged**